here-naksha-lib-psql/src/commonMain/kotlin/naksha/psql/executors/query/WhereClauseBuilder.kt

```diff
@@ -0,0 +1,221 @@
+    }
+
+    private fun whereSpatial() {
+        request.query.spatial?.let { spatialQuery ->
+            if (where.isNotEmpty()) {
+                where.append(" AND ")
+            }
+            where.append("(")
+            whereNestedSpatial(spatialQuery)
+            where.append(")")
+        }
+    }
+
+    private tailrec fun whereNestedSpatial(spatial: ISpatialQuery) {
+        when (spatial) {
+            is SpNot -> {
+                where.append("NOT ")
+                whereNestedSpatial(spatial.query)
+            }
+
+            is SpAnd -> {
+                for (i in spatial.indices) {
+                    val subSpatial = spatial[i]
+                    if (subSpatial != null) {
+                        if (i > 0) where.append("AND ")
+                        whereNestedSpatial(subSpatial)
+                    }
+                }
+            }
+
+            is SpOr -> {
+                for (i in spatial.indices) {
+                    val subSpatial = spatial[i]
+                    if (subSpatial != null) {
+                        if (i > 0) where.append("OR ")
+                        whereNestedSpatial(subSpatial)
+                    }
+                }
+            }
+
+            is SpIntersects -> {
+                // TODO: Add transformations!
+                val twkb = PgUtil.encodeGeometry(
+                    spatial.geometry,
+                    Flags().geoGzipOff().geoEncoding(GeoEncoding.TWKB)
+                )
+                val placeholder = placeholderForArg(twkb, PgType.BYTE_ARRAY)
+                where.append("ST_Intersects(naksha_geometry(${PgColumn.geo}, ${PgColumn.flags}), ST_GeomFromTWKB($placeholder))")
+            }
+
+            else -> throw NakshaException(
+                NakshaError.ILLEGAL_ARGUMENT,
+                "Invalid spatial query found: $spatial"
+            )
+        }
+    }
+
+    private fun whereMetadata() {
+        request.query.metadata?.let { metaQuery ->
+            if (where.isNotEmpty()) {
+                where.append(" AND ")
+            }
+            where.append("(")
+            whereNestedMetadata(request, metaQuery, where)
+            where.append(")")
+        }
+    }
+
+    private tailrec fun whereNestedMetadata(
+        request: ReadFeatures,
+        metaQuery: IMetaQuery,
+        where: StringBuilder
+    ) {
+        when (metaQuery) {
+            is MetaNot -> {
+                where.append(" NOT ")
+                whereNestedMetadata(request, metaQuery.query, where)
+            }
+
+            is MetaAnd -> {
+                metaQuery.filterNotNull().forEachIndexed { index, subQuery ->
+                    if (index > 0) {
+                        where.append(" AND ")
+                    }
+                    whereNestedMetadata(request, subQuery, where)
+                }
+            }
+
+            is MetaOr -> {
+                metaQuery.filterNotNull().forEachIndexed { index, subQuery ->
```

Comment:
I think adding parentheses here `where.append("(")` and `where.append(")")` would be needed, as AND has higher priority than OR, and without parentheses the logic will become flawed.

Reply:
Nicely caught, I will adjust this in other parts as well (not only meta sufers from it)